src/back/wgsl/writer.rs

```diff
@@ -122,23 +213,395 @@ impl<W: Write> Writer<W> {
     /// Ends in a newline
     fn write_struct(
         &mut self,
+        module: &Module,
         name: &str,
         block: bool,
-        _members: &[StructMember],
+        members: &[StructMember],
     ) -> BackendResult {
         if block {
-            self.write_decoration(Decoration::Block)?;
+            self.write_attribute(Attribute::Block)?;
             writeln!(self.out)?;
         }
         write!(self.out, "struct {} {{", name)?;
-        write!(self.out, "}}")?;
+        writeln!(self.out)?;
+        for (_, member) in members.iter().enumerate() {
+            // The indentation is only for readability
+            write!(self.out, "{}", INDENT)?;
+            if let Some(ref binding) = member.binding {
+                self.write_attribute(map_binding_to_attribute(binding))?;
+                write!(self.out, " ")?;
+            }
+            // Write struct member name and type
+            write!(self.out, "{}: ", member.name.as_ref().unwrap())?;
+            self.write_type(module, member.ty)?;
+            write!(self.out, ";")?;
+            writeln!(self.out)?;
+        }
+        write!(self.out, "}};")?;
 
         writeln!(self.out)?;
 
         Ok(())
     }
 
+    /// Helper method used to write non image/sampler types
+    ///
+    /// # Notes
+    /// Adds no trailing or leading whitespace
+    fn write_type(&mut self, module: &Module, ty: Handle<Type>) -> BackendResult {
+        let inner = &module.types[ty].inner;
+        match *inner {
+            TypeInner::Struct { .. } => {
+                // Get the struct name
+                let name = &self.names[&NameKey::Type(ty)];
+                write!(self.out, "{}", name)?;
+                return Ok(());
+            }
+            ref other => self.write_value_type(module, other)?,
+        }
+
+        Ok(())
+    }
+
+    /// Helper method used to write value types
+    ///
+    /// # Notes
+    /// Adds no trailing or leading whitespace
+    ///
+    fn write_value_type(&mut self, _module: &Module, inner: &TypeInner) -> BackendResult {
+        match *inner {
+            TypeInner::Vector { size, kind, width } => write!(
+                self.out,
+                "{}",
+                format!(
+                    "vec{}<{}>",
+                    vector_size_string(size),
+                    scalar_size_string(kind, width),
+                )
+            )?,
+            _ => todo!("write_value_type {:?}", inner),
+        }
+
+        Ok(())
+    }
+    /// Helper method used to write statements
+    ///
+    /// # Notes
+    /// Always adds a newline
+    fn write_stmt(
+        &mut self,
+        module: &Module,
+        stmt: &Statement,
+        func_ctx: &FunctionCtx<'_>,
+        indent: usize,
+    ) -> BackendResult {
+        match *stmt {
+            Statement::Emit(ref range) => {
+                for handle in range.clone() {
+                    let min_ref_count = func_ctx.expressions[handle].bake_ref_count();
```

Comment:
if you notice that there are re-usable parts possible between MSL/GLSL/WGSL, let's make a note somewhere to try moving them out into a common area